File: dynamicProgramming/numberOfSmoothDescentPeriodsOfAStock.py
# https://leetcode.com/problems/number-of-smooth-descent-periods-of-a-stock/

def getDescentPeriods(prices):
            
    #1  faster than 40% other solutions O(n)
    
    # descent=[1]*len(prices) 
    # for i in range(1, len(prices)):
    #     if prices[i-1]-prices[i]==1:
    #         descent[i]+=descent[i-1]   
    # return sum(descent)
        
    #2  Improved above solution space complexity -->O(1)
    # final, temp=1,1
    # for i in range(1, len(prices)):
    #     if prices[i-1]-prices[i]==1:
    #         temp+=1
    #     else:
    #         temp=1
    #     final+=temp
    # return final
    
    #3  Faster than 80%
    # final, temp, prev=0,1,
    # for i in prices:
    #     if prev-i==1:
    #         temp+=1
    #     else:
    #         temp=1
    #     final+=temp
    #     prev=i
    # return final

    #4  Faster than 80% other solutions (CALCULATING FORMULA)
    counter=len(prices)
    leng=1
    for i in range(1,len(prices)):
        if prices[i-1]-prices[i]==1:
            leng+=1

        else:
            counter+=(((leng-1)*leng)//2)
            leng=1
    if leng>1:
        counter+=(((leng-1)*leng)//2 )
    return counter

prices = [3,2,1,4]
print(getDescentPeriods(prices))

# Jiganesh
class Solution(object):
    def getDescentPeriods(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        start = 0
        end=val=1
        
        while end < len(prices):
            if (prices[end-1]-prices[end])==1:
                val+=end-start
            else :
                start=end
            val+=1
            end+=1
                
        return val

File: dynamicProgramming/numberOfSmoothDescentPeriodsOfAStock.java
package dynamicProgramming;

// https://leetcode.com/problems/number-of-smooth-descent-periods-of-a-stock/

public class numberOfSmoothDescentPeriodsOfAStock{
    public static void main(String[] args) {

        SolutionNOSDPOAS solution = new SolutionNOSDPOAS();
        System.out.println(solution.getDescentPeriods(new int [] {3,2,1,4}));
        
    }
}

class SolutionNOSDPOAS{

    // Submitted by @Jiganesh

    // TC : O(N)
    // SC : O(1)
    public long getDescentPeriods(int[] prices) {
        long start=0, val=1;
        for (int i =1 ; i< prices.length; i++){
            if(prices[i-1]-prices[i]==1){
                val += i-start+1;
            }else{
                start=i;
                val+=1;
            }
        }
        return val;        
    }
    
    /* Example


    prices = [3,2,1,4]

    0       1           2           3
    
    3       2           1           4
    [3]     [2]         [1]         [4]
            [3,2]       [2,1]
                        [3,2,1]

    nextCount = previous + end - start + 1
    
    
    */

    
}